### PR TITLE
CRIU resets j.l.VirtualThread.ForkJoinPool.parallelism after restore

### PR DESCRIPTION
**Pull request by JasonFengJ9**
CRIU resets `j.l.VirtualThread.ForkJoinPool.parallelism` after restore

If there is no system property `jdk.virtualThreadScheduler.parallelism` set at VM startup, `j.l.VirtualThread.ForkJoinPool.parallelism` value is reset to match `Runtime.getRuntime().availableProcessors()` after CRIU
restore;
Added `VM_VMHelpers` methods `getStaticFieldObject`/`findinstanceFieldOffset`/`setObjectFieldI32`;
Changed `hashClassTableAt` to `peekClassHashTable`;
Minor refactoring;
Added tests.

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

File: runtime/vm/CRIUHelpers.cpp [hunk 1]
```diff
@@ -200,7 +200,7 @@ juRandomReseed(J9VMThread *currentThread, void *userData, const char **nlsMsgFor
 	IDATA seedOffset = findinstanceFieldOffsetHelper(currentThread, hookRecord->instanceType, "seed", "Ljava/util/concurrent/atomic/AtomicLong;");
 	if (-1 != seedOffset) {
 #define JUCA_ATOMICLONG "java/util/concurrent/atomic/AtomicLong"
-		J9Class *jucaAtomicLongClass = hashClassTableAt(vm->systemClassLoader, (U_8 *)JUCA_ATOMICLONG, LITERAL_STRLEN(JUCA_ATOMICLONG));
+		J9Class *jucaAtomicLongClass = peekClassHashTable(currentThread, vm->systemClassLoader, (U_8 *)JUCA_ATOMICLONG, LITERAL_STRLEN(JUCA_ATOMICLONG));
```
Comment (tajila) [inline runtime/vm/CRIUHelpers.cpp:203]: why is this changed?

File: runtime/vm/CRIUHelpers.cpp [hunk 2]
```diff
@@ -456,13 +456,30 @@ initializeCriuHooks(J9VMThread *currentThread)
 		}
 	}
 
+#if JAVA_SPEC_VERSION >= 20
+	{
+		J9VMSystemProperty *vtParallelism = NULL;
+		PORT_ACCESS_FROM_VMC(currentThread);
+		if (J9SYSPROP_ERROR_NONE != getSystemProperty(vm, "jdk.virtualThreadScheduler.parallelism", &vtParallelism)) {
```
Comment (tajila) [inline runtime/vm/CRIUHelpers.cpp:463]: we need to make sure this is done after processing the restore args

Comment (tajila) [reply runtime/vm/CRIUHelpers.cpp:463]: Since user can set it at startup and on restore so we need to detect both.

File: runtime/vm/CRIUHelpers.cpp [hunk 3]
```diff
@@ -602,6 +619,58 @@ runInternalJVMRestoreHooks(J9VMThread *currentThread, const char **nlsMsgFormat)
...
+		UDATA cpuCount = j9sysinfo_get_number_CPUs_by_type(J9PORT_CPU_TARGET);
+		if (cpuCount < 1) {
+			/* This matches ForkJoinPool default constructor with Runtime.getRuntime().availableProcessors(). */
+			cpuCount = 1;
+		}
+		if (cpuCount != vm->checkpointState.checkpointCPUCount) {
+#define JAVA_LANG_VIRTUALTHREAD "java/lang/VirtualThread"
+			J9Class *jlVirtualThreadClass = peekClassHashTable(currentThread, vm->systemClassLoader, (U_8 *)JAVA_LANG_VIRTUALTHREAD, LITERAL_STRLEN(JAVA_LANG_VIRTUALTHREAD));
+#undef JAVA_LANG_VIRTUALTHREAD
+			void *defaultSchedulerAddress = staticFieldAddress(currentThread, jlVirtualThreadClass, (U_8*)"DEFAULT_SCHEDULER", strlen("DEFAULT_SCHEDULER"), (U_8*)"Ljava/util/concurrent/ForkJoinPool;", strlen("Ljava/util/concurrent/ForkJoinPool;"), NULL, NULL, 0, NULL);
```
Comment (tajila) [inline runtime/vm/CRIUHelpers.cpp:635]: can you put the code that gets the address and does the update of the parallelism field in a helper in VMHelpers, we will need it for other things

File: runtime/vm/CRIUHelpers.cpp (hunk 1 above)
Comment (JasonFengJ9) [reply runtime/vm/CRIUHelpers.cpp:203]: `hashClassTableAt ()` needs to be protected w/ `classTableMutex`, which `peekClassHashTable()` does.

File: runtime/vm/CRIUHelpers.cpp (hunk 2 above)
Comment (JasonFengJ9) [reply runtime/vm/CRIUHelpers.cpp:446]: `jdk.virtualThreadScheduler.parallelism` is read via [private static final ForkJoinPool DEFAULT_SCHEDULER = createDefaultScheduler()](https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/4b0d1648f68805474340a94cd7b3f173bc6c83a4/src/java.base/share/classes/java/lang/VirtualThread.java#L76) at VM startup.
`j.l.VirtualThread.ForkJoinPool.parallelism` won't be affected by the system property specified by the restore option file.
Added a test to confirm this behaviour.

File: runtime/vm/CRIUHelpers.cpp (hunk 3 above)
Comment (JasonFengJ9) [reply runtime/vm/CRIUHelpers.cpp]: Yeah, added `getStaticFieldObject`/`findinstanceFieldOffset`/`setObjectFieldI32` VMHelper methods.

File: runtime/oti/VMHelpers.hpp [hunk 4]
```diff
@@ -2062,6 +2063,87 @@ class VM_VMHelpers
 		return oldState;
 	}
 
+	/**
+	 * Get a static field object within a defining class.
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/oti/VMHelpers.hpp:2078]: Added.

File: runtime/nls/j9vm/j9vm.nls [hunk 5]
```diff
@@ -2151,4 +2151,27 @@ J9NLS_VM_ILLEGAL_THREAD_STATE_UPCALL=The JVM failed to proceed due to the wrong
 J9NLS_VM_ILLEGAL_THREAD_STATE_UPCALL.explanation=An error occurred when the JVM attempted to perform upcall in the trivial downcall
 J9NLS_VM_ILLEGAL_THREAD_STATE_UPCALL.system_action=The JVM will throw a IllegalThreadStateException.
 J9NLS_VM_ILLEGAL_THREAD_STATE_UPCALL.user_response=Ensure the specified linker options for downcall are valid.
+
+J9NLS_VM_CRIU_RESTORE_RESET_VIRTUALTHREAD_FORKJOINPOOL_PARALLELISM_FORKJOINPOOL_CNF=The JVM could not reset j.l.VirtualThread.ForkJoinPool.parallelism due to j.u.c.ForkJoinPool class not found
```
Comment (tajila) [inline runtime/nls/j9vm/j9vm.nls:2155]: we should use the full names, java.lang..., java.util.concurrent...

Comment (JasonFengJ9) [reply runtime/nls/j9vm/j9vm.nls]: Updated in `j9vm.nls`, assuming the abbreviations in those code comments are okay.

**Comment by tajila** [comment]
Im wondering if we should just use vmconstantpool for this. Im trying to think of a downside, but I think the cost to startup is negligeable

**Comment by tajila** [comment]
I mention it here, because with VirtualThreads ForkJoinPool will certainly be used. I guess it may not be with pre-JDK 21 builds

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
jenkins compile win jdk8

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21
